File: leetcode/js/isBTSymmetric.js
// Given a binary tree, check whether it is a mirror of itself(ie, symmetric around its center).

function TreeNode(val) {
  this.val = val;
  this.left = this.right = null;
}

const isSymmetric = root => {
  if (!root || (!root.left && !root.right)) return true;
  return checkSiblings(root.left, root.right);
};

const checkSiblings = (left, right) => {
  if (!left && !right) return true;
  if ((!left && right) || (left && !right)) return false;
  return (
    left.val === right.val &&
    checkSiblings(left.left, right.right) &&
    checkSiblings(right.left, left.right)
  );
};

File: leetcode/ruby/isBTSymmetric.rb
# Given a binary tree, check whether it is a mirror of itself(ie, symmetric around its center).

class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val)
    @val = val
    @left = @right = nil, nil
  end
end

def isSymmetric?(root)
  return true if (!root || (!root.left && !root.right))
  return checkSiblings(root.left, root.right)
end

def checkSiblings(left, right)
  return true if (!left && !right)
  return false if (left.nil? ^ right.nil?)
  return left.val == right.val && checkSiblings(left.left, right.right) && checkSiblings(right.left, left.right)
end
